antarest/storage/business/study_download_utils.py

```diff
@@ -0,0 +1,319 @@
+                        )
+                    else:
+                        StudyDownloader.apply_type_filters(
+                            matrix,
+                            prefix,
+                            year,
+                            type_elm,
+                            elm,
+                            study,
+                            filtered_url,
+                            data,
+                            flt == elm,
+                            lambda x: True,
+                        )
+            else:  # FilterIn, FilterOut
+                flt_out_1, flt_out_2 = (
+                    StudyDownloader.get_filters(data.filterOut)
+                    if data.filterOut
+                    else ("", "")
+                )
+                flt_in_1, flt_in_2 = (
+                    StudyDownloader.get_filters(data.filterIn)
+                    if data.filterIn
+                    else ("", "")
+                )
+
+                first_element_type_condition_in = bool(
+                    re.search(flt_in_1, elm)
+                )
+                first_element_type_condition_out = (
+                    not bool(re.search(flt_out_1, elm)) if flt_out_1 else True
+                )
+                first_element_type_condition = (
+                    first_element_type_condition_in
+                    and first_element_type_condition_out
+                )
+
+                def second_element_type_condition(x: str) -> bool:
+                    cond1 = not (flt_out_2 and bool(re.search(flt_out_2, x)))
+                    cond2 = not (flt_in_2 and not bool(re.search(flt_in_2, x)))
+                    return cond1 and cond2
+
+                StudyDownloader.apply_type_filters(
+                    matrix,
+                    prefix,
+                    year,
+                    type_elm,
+                    elm,
+                    study,
+                    filtered_url,
+                    data,
+                    first_element_type_condition,
+                    second_element_type_condition,
+                )
+
+    @staticmethod
+    def type_output_filter(
+        matrix: MatrixAggregationResult,
+        prefix: str,
+        year: int,
+        config: StudyConfig,
+        study: Study,
+        url: str,
+        data: StudyDownloadDTO,
+    ) -> None:
+        if StudyDownloadType.from_dict(data.type) == StudyDownloadType.AREA:
+            StudyDownloader.select_filter(
+                matrix, prefix, year, config.areas, study, f"{url}/areas", data
+            )
+        elif (
+            StudyDownloadType.from_dict(data.type) == StudyDownloadType.CLUSTER
+        ):
+            StudyDownloader.select_filter(
+                matrix, prefix, year, config.sets, study, f"{url}/areas", data
+            )
+        else:
+            StudyDownloader.select_filter(
+                matrix, prefix, year, config.areas, study, f"{url}/links", data
+            )
```

Comment:
you can extract once the type:
`data_type = StudyDownloadType.from_dict(data.type)` and then pass it to the select_filter method so that it can more easily choose the correct filtering behavior depending on requested type.